devcareer/code-jammers-backend
# #175537319 Build historical facts CRUD APIs

## What does this PR do ##
-  enables user to add historical Fact
- enables user to delete historical Fact
- enables user to edit historical Fact
- enables user to get historical Fact

### Description of tasks to be completed? ###
- create add historicalFact controller
- create add historicalFact route
- create get historicalFact controller
- create get historicalFact route
- create edit historicalFact controller
- create edit historicalFact route
- create delete historicalFact controller
- create delete historicalFact route
- write tests for all controllers and route

### How should this be manually tested ###
- Clone the PR
- Checkout to the branch `feat-historicalFacts-CRUD-API-175537319`
- Install the dependencies using `npm install`
- Create a `.env file` and define the following as follows:
   - `DEV_DATABASE_URL = <development database url>`
   - `TEST_DATABASE_URL = <test database url>`
   - `DATABASE_URL = <production database url>`
- run ```npm run migrate``` to create the database tables
- run ```npm run seed``` to seed the database
- Start the server by running ```npm run start-dev```
- Sign in via `http://localhost:3000/api/v1/users/signin` using a `POST` method with
```
   {
    "email": "ufuoma@gmail.com",
    "password": "12345"
   }
```
and copy the token (without the quotes) from the response body.
- You need to sign in with email ```francis@gmail.com``` because he is an admin and only an admin has the right to add, update and delete historical fact

- paste the token in the authorization header
- make a POST request to the url http://127.0.0.1:3000/api/v1/admin/historical-fact/6003fb36-5112-463e-a1f9-c8944e72412f with request body
```
{
"location": "Lekki",
  "about": "Write anything about Lekki here",
  "gallery": "put gallery url here"
 }
```
- check your database, the historical Fact should be added

- make a GET request to url http://127.0.0.1:3000/api/v1/historical-fact 
- it should return all historical facts in the database

 make a GET request to url http://127.0.0.1:3000/api/v1/historical-fact/id
- it should return the historical Fact associated with the id you put

- make a PATCH request to url http://127.0.0.1:3000/api/v1/admin/historical-fact/id
 with request body 
```
 {
"location": "new location if any",
 "about": "Update the history if you have a reason to",
"gallery": "put new gallery url here if any"
 }
```
- it should update the historical Fact with the related id

- make a delete request to url http://127.0.0.1:3000/api/v1/admin/historical-fact/id
- it should delete that historical Fact with the related id

### Link to architecture


### What are the relevant pivotal tracker stories?
[ #175537319](https://www.pivotaltracker.com/story/show/175537319)
